File: 1966-frequency-of-the-most-frequent-element/1966-frequency-of-the-most-frequent-element.cpp
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end()); // Sort the array → O(N log N)
        long long l = 0, total = 0, maxFreq = 0;

        for (long long r = 0; r < nums.size(); r++) {
            total += nums[r];  // Expand the window

            // Instead of recalculating everything, only adjust total when necessary
            if ((r - l + 1) * nums[r] - total > k) {
                total -= nums[l]; // Remove smallest element in the window
                l++; // Shrink the window
            }

            maxFreq = max(maxFreq, r - l + 1);
        }
        return maxFreq;
    }
};

File: 1966-frequency-of-the-most-frequent-element/1966-frequency-of-the-most-frequent-element.py
class Solution(object):
    def maxFrequency(self,nums,k):
        nums.sort()
        l=0
        total=0
        maxFreq=0
        for r in range(len(nums)):
            total+=nums[r]
            while (r-l+1)*nums[r]-total>k:
                total-=nums[l]
                l+=1
            maxFreq=max(maxFreq,r-l+1)
        return maxFreq